graph/src/components/ethereum/adapter.rs

```diff
@@ -84,9 +86,50 @@ impl From<Error> for EthereumAdapterError {
     }
 }
 
-#[derive(Clone, Debug)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Ord, PartialOrd, Hash)]
+enum LogFilterNode {
+    Contract(Address),
+    Event(EventSig),
+}
+
+/// Corresponds to an `eth_getLogs` call.
+#[derive(Clone)]
+pub struct EthGetLogsFilter {
+    pub contracts: Vec<Address>,
+    pub event_sigs: Vec<EventSig>,
+}
+
+impl fmt::Display for EthGetLogsFilter {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.contracts.len() == 1 {
+            write!(
+                f,
+                "contract {}, {} events",
+                self.contracts[0],
+                self.event_sigs.len()
+            )
+        } else if self.event_sigs.len() == 1 {
+            write!(
+                f,
+                "event {}, {} contracts",
+                self.event_sigs[0],
+                self.contracts.len()
+            )
+        } else {
+            write!(f, "unreachable")
+        }
+    }
+}
+
+#[derive(Clone, Debug, Default)]
 pub struct EthereumLogFilter {
-    pub contract_address_and_event_sig_pairs: HashSet<(Option<Address>, H256)>,
+    /// Log filters can be represented as a bipartite graph between contracts and events. An edge
+    /// exists between a contract and an event if a data source for the contract has a trigger for
+    /// the event.
+    contracts_and_events_graph: GraphMap<LogFilterNode, (), petgraph::Undirected>,
+
+    // Event sigs with no associated address, matching on all addresses.
+    wildcard_events: HashSet<EventSig>,
```

Comment:
I wonder if these could be represented as another `LogFilterNode` variant to avoid the special casing. It might even make the filter splitting algorithm more efficient?

Reply:
I preferred to keep this separate because imo it makes the code more obvious and readable.

I considered putting it into the graph but we'd also need a special contract vertex representing all contracts. There might be an optimization there if a wildcard event also appears tied to a contract, but the ideal behaviour is unclear. Wildcards are a niche feature anyhow, people usually want to control what contract is emitting an event so that random contracts can't put data into their subgraph.

Reply:
That's true. It was just an idea.